File: pnia/src/main/java/io/github/jlmc/reactive/api/v1/PniaController.java
package io.github.jlmc.reactive.api.v1;

import io.github.jlmc.reactive.domain.usecases.GetPhoneNumberAggregation;
import lombok.extern.slf4j.Slf4j;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Map;

import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;

@Validated
@RestController
@Slf4j
@RequestMapping(path = "aggregate")

public class PniaController {

    private final GetPhoneNumberAggregation aggregation;

    public PniaController(GetPhoneNumberAggregation aggregation) {
        this.aggregation = aggregation;
    }

    @PostMapping(
            produces = APPLICATION_JSON_VALUE,
            consumes = APPLICATION_JSON_VALUE

    )
    public Mono<Map<String, Map<String, Long>>> createAggregation(
            @RequestBody @Validated List<String> phoneNumbers) {

        return aggregation.execute(phoneNumbers);
    }
}

File: spring-boot-tipps-reactive/pnia-kotlin/src/main/kotlin/io/github/jlmc/pniakotlin/api/PniaController.kt
package io.github.jlmc.pniakotlin.api

import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.bind.annotation.RequestMapping
import io.github.jlmc.pniakotlin.domain.interactors.GetPhoneNumberAggregation
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import reactor.core.publisher.Mono

@Validated
@RestController
@RequestMapping(path = ["aggregate"])
class PniaController(private val aggregation: GetPhoneNumberAggregation) {

    @PostMapping(produces = [MediaType.APPLICATION_JSON_VALUE], consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun createAggregation(
        @RequestBody @Validated phoneNumbers: List<String?>
    ): Mono<Map<String, Map<String, Long>>> {
        return aggregation.execute(phoneNumbers)
    }
}
